pkg/tools/clientmap/refcount.go

```diff
@@ -17,68 +17,105 @@
 package clientmap
 
 import (
-	"context"
+	"sync"
 	"sync/atomic"
 
-	"github.com/golang/protobuf/ptypes/empty"
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
-	"google.golang.org/grpc"
 )
 
 type refcountClient struct {
-	count  int32
-	client networkservice.NetworkServiceClient
-}
-
-func (r *refcountClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
-	return r.client.Request(ctx, request)
-}
-
-func (r *refcountClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
-	return r.client.Close(ctx, conn)
+	count int32
+	networkservice.NetworkServiceClient
 }
 
 // RefcountMap - clientmap.Map wrapped with refcounting.  Each Store,Load,LoadOrStore increments the count,
 //               each Delete decrements the count.  When the count is zero, Delete deletes.
 type RefcountMap struct {
-	Map
+	m    Map
+	lock sync.RWMutex
```

Comment:
Why do we need a lock here?

Reply:
For `Delete`, `LoadAndDelete` we need to be sure that no one should update `entry.refcount` when we are going to delete an entry.

Reply:
@Bolodya1997 Then I guess we should not to use sync.Map it looks like an over synchronization

Reply:
Yep, it was too slow in comparison with simple `map + sync.Mutex`. But if we put `sync.RWMutex` in every `refcountClient` instead of keeping single lock for all map, it becomes significantly faster.
```
BenchmarkRefcountMap
BenchmarkRefcountMap-4       	 6499250	       212 ns/op
BenchmarkSyncRefcountMap
BenchmarkSyncRefcountMap-4   	 4078995	       307 ns/op
```
Reworked `RefcountMap` + added bench tests.